File: Vm.hs
module Vm
  ( execVmProgram
  , debugExecVmProgram
  , Stack
  , Memory
  , VmState
  )
where

import qualified Lir
import qualified Intr

import qualified Data.Foldable
import qualified Data.Map.Strict               as M
import           Control.Monad                  ( unless )
import           Control.Monad.Reader
import           Control.Monad.State

import qualified Debug.Trace as Debug

type Stack = [Lir.Value]
type Memory = [M.Map String Lir.Value]

data VmState = VmState
  { memory   :: Memory
  , stack    :: Stack
  , retAddrs :: [Lir.InstrAddr]
  , pc       :: Lir.InstrAddr
  , instrs   :: [Lir.Instr]
  , running  :: Bool
  }
  deriving Show

type VmProgram = StateT VmState IO

pop :: VmProgram Lir.Value
pop = do
  state <- get
  case stack state of
    (top : stack') -> do
      put $ state { stack = stack' }
      return top
    [] -> error $ "\n\tVM Stack Underflow!\n\tstate = " ++ show state ++ "\n"

popInt :: VmProgram Int
popInt = do
  val <- pop
  let Lir.VInt x = val
  return x

popBool :: VmProgram Bool
popBool = do
  val <- pop
  let Lir.VBool x = val
  return x

popText :: VmProgram String
popText = do
  val <- pop
  let Lir.VText x = val
  return x

popInstrAddr :: VmProgram Lir.InstrAddr
popInstrAddr = do
  val <- pop
  case val of
    Lir.VInstrAddr x -> return x
    other -> error $ "Interal Vm Error: Expected VInstrAddr, got: " ++ show other

push :: Lir.Value -> VmProgram ()
push value = do
  state <- get
  put $ state { stack = value : stack state }

load :: String -> VmProgram ()
load var = do
  state <- get
  push $ findInFrames $ memory state
    where
      findInFrames [] = error ("Unbound variable \"" ++ var ++ "\"")
      findInFrames (frame:frames) =
        case M.lookup var frame of
             Just val -> val
             Nothing  -> findInFrames frames

store :: String -> VmProgram ()
store var = do
  val   <- pop
  state <- get
  let (frame : parentFrames) = memory state
  let frame'                 = M.insert var val frame
  put $ state { memory = frame' : parentFrames }

setPc :: Lir.InstrAddr -> VmProgram ()
setPc pc' = do
  modify $ \state -> state { pc = pc' }

incrPc :: VmProgram ()
incrPc = do
  oldPc <- gets pc
  setPc (oldPc + 1)

pushRetAddr :: Lir.InstrAddr -> VmProgram ()
pushRetAddr retAddr = do
  retAddrs_ <- gets retAddrs
  modify $ \st -> st { retAddrs = (retAddr : retAddrs_) }

popRetAddr :: VmProgram Lir.InstrAddr
popRetAddr = do
  retAddrs_ <- gets retAddrs
  case retAddrs_ of
    retAddr:retAddrs' -> do
      modify $ \st -> st { retAddrs = retAddrs' }
      return retAddr
    [] -> error "Vm Return Stack Underflow!"

pushNewFrame :: VmProgram ()
pushNewFrame = do
  oldMem <- gets memory
  let newMem = M.empty : oldMem
  modify $ \state -> state { memory = newMem }

popMemFrame :: VmProgram ()
popMemFrame = do
  oldMem <- gets memory
  let (_ : newMem) = oldMem
  modify $ \state -> state { memory = newMem }

stepIntBinOp :: (Int -> Int -> Int) -> VmProgram ()
stepIntBinOp op = do
  x <- popInt
  y <- popInt
  push $ Lir.VInt $ x `op` y

stepSameTypeBoolBinOp :: (Lir.Value -> Lir.Value -> Bool) -> VmProgram ()
stepSameTypeBoolBinOp op = do
  x <- pop
  y <- pop
  push $ Lir.VBool $ x `op` y

stepIntBoolBinOp :: (Int -> Int -> Bool) -> VmProgram ()
stepIntBoolBinOp op = do
  x <- popInt
  y <- popInt
  push $ Lir.VBool $ x `op` y

stepBoolBinOp :: (Bool -> Bool -> Bool) -> VmProgram ()
stepBoolBinOp op = do
  x <- popBool
  y <- popBool
  push $ Lir.VBool $ x `op` y

stepVm :: Lir.Instr -> VmProgram ()
stepVm instr = do
  step
  incrPc
 where
  step = case instr of
    Lir.Load var -> do
      load var
    Lir.Store var -> do
      store var
    Lir.Const val -> do
      push val
    Lir.Dup -> do
      tmp <- pop
      push tmp
      push tmp
    -- [a, b, ...] -> [b, a, b, ...]
    Lir.Over -> do
      a <- pop
      b <- pop
      push b
      push a
      push b
    -- [a, b, c, ...] -> [c, a, b, ...]
    Lir.Rot -> do
      a <- pop
      b <- pop
      c <- pop
      push c
      push a
      push b
    -- [a, b, ...] -> [b, a, ...]
    Lir.Swap -> do
      a <- pop
      b <- pop
      push a
      push b
    Lir.Pop -> do
      _ <- pop
      return ()
    Lir.Add -> stepIntBinOp (+)
    Lir.Sub -> stepIntBinOp (-)
    Lir.Mul -> stepIntBinOp (*)
    Lir.Div -> stepIntBinOp div
    Lir.Neg -> do
      i <- popInt
      push (Lir.VInt (-i))
    Lir.Gt  -> stepIntBoolBinOp (>)
    Lir.Lt  -> stepIntBoolBinOp (<)
    Lir.And -> stepBoolBinOp (&&)
    Lir.Or  -> stepBoolBinOp (||)
    Lir.Not -> do
      b <- popBool
      push (Lir.VBool (not b))
    Lir.Concat -> do
      a <- popText
      b <- popText
      push $ Lir.VText $ a ++ b
    Lir.Jmp        idx -> setPc (idx - 1)
    Lir.JmpIfFalse idx -> do
      b <- popBool
      unless b $
                 -- we always incr pc, so goto idx - 1
                 setPc (idx - 1)
    Lir.Nop -> return ()
    Lir.Intrinsic intr -> runIntrinsic intr
    -- Assume stack is set up properly beforehand.
    -- Stack should look like this:
    --  | <the function's InstrAddr>   <-- TOS
    --  | <arg n>
    --  | <arg n-1>
    --  | ...
    --  | <arg 2>
    --  | <arg 1>
    -- After running Lir.Call, stack should like like this:
    --  | <return address>
    --  | <arg n>
    --  | <arg n-1>
    --  | ...
    --  | <arg 2>
    --  | <arg 1>
    Lir.Call argC -> do
      fnAddr <- popInstrAddr -- Get the function's entry address.
      callDirect fnAddr argC

    Lir.CallDirect fnAddr argC -> do
      callDirect fnAddr argC

    Lir.Ret -> do
      retAddr <- popRetAddr
      popMemFrame
      setPc (retAddr - 1)
      return ()

callDirect :: Lir.InstrAddr -> Int -> VmProgram ()
callDirect fnAddr argC = do
  -- Store the return address above all the args.
  pc_ <- gets pc
  let retAddr = pc_ + 1 -- Return to the NEXT instr.
  pushRetAddr retAddr

  -- Perform the jump.
  setPc (fnAddr - 1)
  pushNewFrame
  return ()

runIntrinsic :: Intr.Intrinsic -> VmProgram ()
runIntrinsic op = case op of
  Intr.EqInt -> do
    x <- popInt
    y <- popInt
    push $ Lir.VBool $ x == y
  Intr.EqBool -> do
    x <- popBool
    y <- popBool
    push $ Lir.VBool $ x == y
  Intr.EqText -> do
    x <- popText
    y <- popText
    push $ Lir.VBool $ x == y
  Intr.DbgInt -> do
    x <- popInt
    lift $ Lir.dbgValue $ Lir.VInt x
  Intr.DbgBool -> do
    x <- popBool
    lift $ Lir.dbgValue $ Lir.VBool x
  Intr.DbgText -> do
    x <- popText
    lift $ Lir.dbgValue $ Lir.VText x
  Intr.Puts -> do
    x <- popText
    lift $ putStrLn x
  Intr.Here pos -> do
    lift $ putStrLn ("intr.here[] at " ++ show pos)
  Intr.Exit -> do
    modify $ \state -> state { running = False }

nth :: (Ord i, Num i) => [a] -> i -> Maybe a
nth _ n | n < 0 = Nothing
nth []       n  = Nothing
nth (x : xs) 0  = Just x
nth (_ : xs) n  = nth xs (n - 1)

debugStepProgram :: Lir.Instr -> VmProgram ()
debugStepProgram instr = do
  instrs_   <- gets instrs
  pc_       <- gets pc
  retAddrs_ <- gets retAddrs
  stack_    <- gets stack
  memory_   <- gets memory
  let Lir.InstrAddr pcNum = pc_
  lift $ putStrLn $ "\tInstr #" ++ show pcNum ++ ": " ++ show instr
  lift $ putStr "\tStack: "
  lift $ print stack_
  lift $ putStr "\tReturn Addresses: "
  lift $ print retAddrs_
  lift $ putStr "\tMemory: "
  lift $ print memory_
  stepVm instr

debugRunProgram :: VmProgram ()
debugRunProgram = do
  pc_     <- gets pc
  instrs_ <- gets instrs
  vmState <- get
  when (running vmState) $
    case nth instrs_ pc_ of
      Just instr -> do
        debugStepProgram instr
        lift $ putStr "\tPress ENTER to step forward...\n"
        input <- lift $ getLine
        if input == "q" || input == "Q" || input == "quit"
          then return ()
          else debugRunProgram
      Nothing -> error $ "VM: Illegal instruction address: " ++ show pc_

runProgram :: VmProgram ()
runProgram = do
  pc_     <- gets pc
  instrs_ <- gets instrs
  vmState <- get
  when (running vmState) $
    case nth instrs_ pc_ of
      Just instr -> do
        stepVm instr
        runProgram
      Nothing -> error $ "VM: Illegal instruction address: " ++ show pc_

initState :: [Lir.Instr] -> VmState
initState instrs =
  VmState { memory = [M.empty]
          , stack = []
          , pc = 0
          , retAddrs = []
          , instrs = instrs
          , running = True
          }

execVmProgram :: [Lir.Instr] -> IO VmState
execVmProgram program = execStateT runProgram (initState program)

debugExecVmProgram :: [Lir.Instr] -> IO VmState
debugExecVmProgram program = execStateT debugRunProgram (initState program)

File: vm-rs/src/vm.rs
use std::{
    collections::HashMap,
    io::{self, Write}
};

use crate::instr::{Ident, Instr, InstrAddr, Intrinsic, Value};

#[derive(Debug)]
pub struct Heap {
    mem: Vec<Value>,
}

impl Heap {
    fn new() -> Self {
        Self { mem: Vec::new() }
    }

    fn alloc(&mut self, slots: usize) -> usize {
        let idx = self.mem.len();
        // Extend and fill the new part with zeros.
        self.mem.resize_with(self.mem.len() + slots, || Value::VInt(0));
        idx
    }
}

#[derive(Debug)]
pub struct Vm {
    memory: Vec<HashMap<String, Value>>,
    stack: Vec<Value>,
    ret_addrs: Vec<InstrAddr>,
    pc: InstrAddr,
    running: bool,
}

impl Vm {
    pub fn new() -> Self {
        Self {
            memory: Vec::new(),
            stack: Vec::new(),
            ret_addrs: Vec::new(),
            pc: InstrAddr(0),
            running: true,
        }
    }

    fn push(&mut self, x: Value) {
        self.stack.push(x);
    }

    fn pop(&mut self) -> Value {
        self.stack.pop().expect("Vm: Stack underflow!")
    }

    fn pop_bool(&mut self) -> bool {
        match self.pop() {
            Value::VBool(x) => x,
            x => panic!("Expected VBool at TOS, got {:?}!", x),
        }
    }

    fn pop_int(&mut self) -> i64 {
        match self.pop() {
            Value::VInt(x) => x,
            x => panic!("Expected VInt at TOS, got {:?}!", x),
        }
    }

    fn pop_string(&mut self) -> String {
        match self.pop() {
            Value::VString(x) => x,
            x => panic!("Expected VString at TOS, got {:?}!", x),
        }
    }

    fn pop_instr_addr(&mut self) -> InstrAddr {
        match self.pop() {
            Value::VInstrAddr(x) => x,
            x => panic!("Expected VInstrAddr at TOS, got {:?}!", x),
        }
    }

    fn pop_ptr(&mut self) -> usize {
        match self.pop() {
            Value::VPtr(x) => x,
            x => panic!("Expected VPtr at TOS, got {:?}!", x),
        }
    }

    fn load(&mut self, var: &Ident) {
        for frame in self.memory.iter().rev() {
            if let Some(v) = frame.get(var) {
                let v = v.clone();
                self.push(v);
                return;
            }
        }
        panic!("Unbound variable `{}`!", var);
    }

    fn store(&mut self, var: &Ident) {
        let val = self.pop();
        let frame = self
            .memory
            .last_mut()
            .expect("Should be at least one frame at this point");
        frame.insert(var.clone(), val);
    }

    fn incr_pc(&mut self) {
        let InstrAddr(pc) = self.pc;
        self.pc = InstrAddr(pc + 1);
    }

    fn pop_ret_addr(&mut self) -> InstrAddr {
        self.ret_addrs.pop().expect("Vm Return Stack Underflow!")
    }

    fn push_new_frame(&mut self) {
        self.memory.push(HashMap::new());
    }

    fn pop_mem_frame(&mut self) {
        self.memory.pop().expect("Memory Frame Stack Underflow!");
    }

    fn tos(&self) -> &Value {
        self.stack.last().expect("Stack underflow!")
    }

    fn run_intrinsic(&mut self, intr: &Intrinsic) {
        match intr {

            Intrinsic::EqInt => {
                let x = self.pop_int();
                let y = self.pop_int();
                self.push(Value::VBool(x == y))
            }

            Intrinsic::EqBool => {
                let x = self.pop_bool();
                let y = self.pop_bool();
                self.push(Value::VBool(x == y))
            }

            Intrinsic::EqText => {
                let x = self.pop_string();
                let y = self.pop_string();
                self.push(Value::VBool(x == y))
            }

            Intrinsic::DbgInt => println!("{:?}", self.pop_int()),

            Intrinsic::DbgBool => println!("{:?}", self.pop_bool()),

            Intrinsic::DbgText => println!("{:?}", self.pop_string()),

            Intrinsic::Puts => println!("{}", self.pop()),

            Intrinsic::Here(loc) => println!("{}", loc),

            Intrinsic::Exit => self.running = false,
        }
    }

    fn call_direct(&mut self, InstrAddr(fn_addr): InstrAddr, _argc: u8) {
        // Store the return address above all the args.
        let InstrAddr(pc) = self.pc;
        self.ret_addrs.push(InstrAddr(pc + 1)); // Return to the NEXT instr.

        // Perform the jump.
        self.pc = InstrAddr(fn_addr - 1);
        self.push_new_frame();
    }

    fn step(&mut self, code: &[Instr], heap: &mut Heap, debug: bool) {
        let instr = &code[self.pc.0];

        if debug {
            eprintln!("--------------------");
            eprintln!("instr {} = {:?}", self.pc, instr);
            eprintln!("stack = {:?}", self.stack);

            eprint!(">>> ");
            io::stdout().flush().unwrap();
            let mut buf = String::new();
            io::stdin().read_line(&mut buf).unwrap();
            if buf.to_uppercase().trim() == "Q" {
                self.running = false;
                eprintln!("EXITING...");
            }
        }

        match instr {
            Instr::Load(var) => self.load(var),

            Instr::Store(var) => self.store(var),

            Instr::Const(val) => self.push(val.clone()),

            Instr::Dup => {
                let tos = self.tos().clone();
                self.push(tos);
            }

            // [a, b, ...] -> [b, a, b, ...]
            Instr::Over => {
                let a = self.pop();
                let b = self.pop();
                self.push(b.clone());
                self.push(a);
                self.push(b);
            }

            // [a, b, c, ...] -> [c, a, b, ...]
            Instr::Rot => {
                let a = self.pop();
                let b = self.pop();
                let c = self.pop();
                self.push(c);
                self.push(a);
                self.push(b);
            }

            // [a, b, ...] -> [b, a, ...]
            Instr::Swap => {
                let a = self.pop();
                let b = self.pop();
                self.push(b);
                self.push(a);
            }

            Instr::Pop => {
                let _ = self.pop();
            }

            Instr::Add => {
                let x = self.pop_int();
                let y = self.pop_int();
                self.push(Value::VInt(x + y));
            }

            Instr::Sub => {
                let x = self.pop_int();
                let y = self.pop_int();
                self.push(Value::VInt(x - y));
            }

            Instr::Mul => {
                let x = self.pop_int();
                let y = self.pop_int();
                self.push(Value::VInt(x * y));
            }

            Instr::Div => {
                let x = self.pop_int();
                let y = self.pop_int();
                self.push(Value::VInt(x / y));
            }

            Instr::Neg => {
                let x = self.pop_int();
                self.push(Value::VInt(-x));
            }

            Instr::And => {
                let x = self.pop_bool();
                let y = self.pop_bool();
                self.push(Value::VBool(x && y));
            }

            Instr::Or => {
                let x = self.pop_bool();
                let y = self.pop_bool();
                self.push(Value::VBool(x || y));
            }

            Instr::Not => {
                let x = self.pop_bool();
                self.push(Value::VBool(!x));
            }

            Instr::Gt => {
                let x = self.pop_int();
                let y = self.pop_int();
                self.push(Value::VBool(x > y));
            }

            Instr::Lt => {
                let x = self.pop_int();
                let y = self.pop_int();
                self.push(Value::VBool(x < y));
            }

            Instr::Concat => {
                let x = self.pop_string();
                let y = self.pop_string();
                self.push(Value::VString(x + &y));
            }

            Instr::Alloc(n) => {
                let idx = heap.alloc(*n);
                self.push(Value::VPtr(idx));
            }

            // BEFORE:
            // TOS -> [val: Value]
            //        [buf_start: Value::VPtr]
            //        [...]
            // AFTER:
            // TOS -> [buf_start: Value::VPtr]
            //        [...]
            Instr::MemWriteDirect(offset) => {
                let val = self.pop();
                let buf_start = self.pop_ptr();
                heap.mem[buf_start + offset] = val;
                self.push(Value::VPtr(buf_start));
            }

            Instr::MemReadDirect(offset) => {
                let buf_start = self.pop_ptr();
                let val = heap.mem[buf_start + offset].clone();
                self.push(val);
            }

            Instr::Nop => {}

            Instr::JmpIfFalse(InstrAddr(addr)) => {
                let cond = self.pop_bool();
                if !cond {
                    // We always incr pc, so goto idx - 1.
                    self.pc = InstrAddr(addr - 1)
                }
            }

            Instr::Jmp(InstrAddr(addr)) => {
                // We always incr pc, so goto idx - 1.
                self.pc = InstrAddr(addr - 1);
            }

            Instr::Intrinsic(intr) => self.run_intrinsic(intr),

            // Assume stack is set up properly beforehand.
            // Stack should look like this:
            //  | <the function's InstrAddr>   <-- TOS
            //  | <arg n>
            //  | <arg n-1>
            //  | ...
            //  | <arg 2>
            //  | <arg 1>
            // After running Lir.Call, stack should like like this:
            //  | <return address>
            //  | <arg n>
            //  | <arg n-1>
            //  | ...
            //  | <arg 2>
            //  | <arg 1>
            // Lir.Call argC -> do
            //   fnAddr <- popInstrAddr -- Get the function's entry address.
            //   callDirect fnAddr argC
            Instr::Call(argc) => {
                let fn_addr = self.pop_instr_addr();
                self.call_direct(fn_addr, *argc);
            }

            Instr::CallDirect(fn_addr, argc) => {
                self.call_direct(*fn_addr, *argc);
            }

            Instr::Ret => {
                let InstrAddr(ret_addr) = self.pop_ret_addr();
                self.pop_mem_frame();
                self.pc = InstrAddr(ret_addr - 1); // We always incr pc, so sub 1.
            }
        }
    }

    pub fn exec(&mut self, code: &[Instr], debug: bool) {
        let mut heap = Heap::new();
        while self.running {
            self.step(code, &mut heap, debug);
            self.incr_pc();
        }
    }
}
